### PR TITLE
[c] Allow to create slots and bones on fly

### PR DESCRIPTION
**Pull request by Darky-Lucera**
Allow to create slots and bones on fly, get attachments by name, and set attachments from other skins

Added functions:
- Skeleton: spSkeleton_addBone, spSkeleton_addSlot, spSkeleton_getAttachmentByName, spSkeleton_setAttachment2, spSkeleton_setAttachmentFromSkin
- Skin: spSkin_getAttachment2

### CODE REVIEW

**Comment by badlogic** [comment]
This is a good addition and should probably be included in all runtimes. @NathanSweet will have to give some input on this.

File: spine-c/include/spine/Skeleton.h [hunk 1]
```diff
@@ -96,31 +98,43 @@ spAttachment* spSkeleton_getAttachmentForSlotName (const spSkeleton* self, const
...
+#define Skeleton_setSkinByName(...)				spSkeleton_setSkinByName(__VA_ARGS__)
+#define Skeleton_getAttachmentForSlotName(...)	spSkeleton_getAttachmentForSlotName(__VA_ARGS__)
 #define Skeleton_getAttachmentForSlotIndex(...) spSkeleton_getAttachmentForSlotIndex(__VA_ARGS__)
-#define Skeleton_setAttachment(...) spSkeleton_setAttachment(__VA_ARGS__)
-#define Skeleton_update(...) spSkeleton_update(__VA_ARGS__)
+#define Skeleton_getAttachmentByName(...)		spSkeleton_getAttachmentByName(__VA_ARGS__)
+#define Skeleton_setAttachment(...)				spSkeleton_setAttachment(__VA_ARGS__)
+#define Skeleton_setAttachment2(...)			spSkeleton_setAttachment2(__VA_ARGS__)
+#define Skeleton_setAttachmentFromSkin(...)		spSkeleton_setAttachmentFromSkin(__VA_ARGS__)
+#define Skeleton_update(...)					spSkeleton_update(__VA_ARGS__)
```
Comment (NathanSweet) [inline spine-c/include/spine/Skeleton.h]: No need to change formatting.

File: spine-c/src/spine/Skeleton.c [hunk 2]
```diff
@@ -234,6 +234,19 @@ void spSkeleton_setSlotsToSetupPose (const spSkeleton* self) {
...
+	self->bones = (spBone **) realloc(self->bones, sizeof(spBone *) * (self->bonesCount + 1));
+	self->bones[self->bonesCount] = bone;
+
+	self->data->bones = (spBoneData **) realloc(self->data->bones, sizeof(spBoneData *) * (self->bonesCount + 1));
+	self->data->bones[self->bonesCount] = bone->data;
+
+	++self->bonesCount;
+	++self->data->bonesCount;
+
+	spSkeleton_updateCache(self);
```
Comment (NathanSweet) [inline spine-c/src/spine/Skeleton.c]: Expensive, should not be called every time a bone is added. Require user to call it.

File: spine-c/src/spine/Skeleton.c [hunk 3]
```diff
@@ -248,6 +261,54 @@ int spSkeleton_findBoneIndex (const spSkeleton* self, const char* boneName) {
...
+				numFrames = drawOrderTL->framesCount;
+				for(f=0; f<numFrames; ++f)
+				{
+					drawOrderTL->drawOrders[f] = (int *) realloc(CONST_CAST(int*, drawOrderTL->drawOrders[f]), sizeof(int) * self->slotsCount);
+					CONST_CAST(int*, drawOrderTL->drawOrders[f])[self->slotsCount-1] = self->slotsCount-1;
+				}
+			}
+		}
+	}
+}
```
Comment (NathanSweet) [inline spine-c/src/spine/Skeleton.c]: The add functions aren't really needed for other runtimes, since other languages expose a list that can be mutated. Fixing up draw order timelines is different though. To merge this we'd need the functionality in all runtimes.

File: spine-c/src/spine/Skeleton.c [hunk 4]
```diff
@@ -311,6 +372,28 @@ spAttachment* spSkeleton_getAttachmentForSlotIndex (const spSkeleton* self, int
...
+		if (attachment) return attachment;
+	}
+	if (self->data->defaultSkin) {
+		spAttachment *attachment = spSkin_getAttachment2(self->data->defaultSkin, attachmentName);
+		if (attachment) return attachment;
+	}
+	return 0;
+}
+
+spAttachment* spSkeleton_getAttachmentFromSkin (const spSkeleton* self, const spSkin *skin, const char* attachmentName)
```
Comment (NathanSweet) [inline spine-c/src/spine/Skeleton.c]: Why is this needed?

**Comment by NathanSweet** [comment]
Added a few comments. It's ok for the most part, with a little fix up. Having `add*` methods for spine-c is a bit easier for users. Other runtimes exposing a list and adding can be done to that. As mentioned in comments above, fixing up the draw order timelines is a convenience that could be done in all runtimes.

File: spine-c/src/spine/Skeleton.c (hunk 4 above)
Comment (Darky-Lucera) [inline spine-c/src/spine/Skeleton.c]: It is handy and useful to get an attachment by name

Comment (NathanSweet) [inline spine-c/src/spine/Skeleton.c]: Sure, but this seems like it would be a method on `spSkin`.

Comment (Darky-Lucera) [inline spine-c/src/spine/Skeleton.c]: Yes, maybe spSkin is a better place

**Comment by Darky-Lucera** [comment]
I only needed these features and pulled the request because I thought that they could be useful and handy for other people. Change whatever you want. It only was a first attempt to do it.

^_^

Best regards!
